From 73711b24d374fe3557a563142cb1a3d750b6305b Mon Sep 17 00:00:00 2001
From: Evan Chen <evanxinchen@google.com>
Date: Tue, 19 Aug 2025 17:33:21 +0000
Subject: [PATCH 08/55] Make sure to disassociate all the packages with shared
 UID

Test: cts
Bug: 427206637
Flag: Flag: EXEMPT bugfix
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:5d85254e025ceba2c89ab03427bd1e2f8b68cc45
Merged-In: I3cccad4e85e77916cff020eb7087ac2c355e3e23
Change-Id: I3cccad4e85e77916cff020eb7087ac2c355e3e23
---
 .../association/DisassociationProcessor.java   | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
index 81bfe5fc6..7fbf00e05 100644
--- a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
+++ b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
@@ -34,6 +34,7 @@ import android.companion.AssociationInfo;
 import android.companion.Flags;
 import android.content.Context;
 import android.content.Intent;
+import android.content.pm.PackageManager;
 import android.content.pm.PackageManagerInternal;
 import android.content.pm.ResolveInfo;
 import android.os.Binder;
@@ -75,6 +76,8 @@ public class DisassociationProcessor {
     @NonNull
     private final PackageManagerInternal mPackageManagerInternal;
     @NonNull
+    private final PackageManager mPackageManager;
+    @NonNull
     private final DevicePresenceProcessor mDevicePresenceMonitor;
     @NonNull
     private final SystemDataTransferRequestStore mSystemDataTransferRequestStore;
@@ -106,6 +109,7 @@ public class DisassociationProcessor {
         mSystemDataTransferRequestStore = systemDataTransferRequestStore;
         mTransportManager = companionTransportManager;
         mNotificationManager = notificationManager;
+        mPackageManager = context.getPackageManager();
     }
 
     /**
@@ -290,16 +294,20 @@ public class DisassociationProcessor {
                 return;
             }
 
-            final String packageName = mPackageManagerInternal.getNameForUid(uid);
-            if (packageName == null) {
+            // A UID can be shared by multiple packages if android:sharedUserId is used.
+            // We must get all packages for the UID to ensure we find the correct one.
+            final String[] packageNames = mPackageManager.getPackagesForUid(uid);
+            if (packageNames == null || packageNames.length == 0) {
                 // Not interested in this uid.
                 return;
             }
 
             int userId = UserHandle.getUserId(uid);
-            for (AssociationInfo association : mAssociationStore.getRevokedAssociations(userId,
-                    packageName)) {
-                disassociate(association.getId(), REASON_REVOKED);
+            for (String packageName : packageNames) {
+                for (AssociationInfo association : mAssociationStore.getRevokedAssociations(userId,
+                        packageName)) {
+                    disassociate(association.getId(), REASON_REVOKED);
+                }
             }
 
             if (mAssociationStore.getRevokedAssociations().isEmpty()) {
-- 
2.52.0

